File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
/////// Same Approach as Longest K unique characters substring /////////

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int i = 0, j = 0, n = fruits.size();
        unordered_map<int, int> mp;
        int res = 1;
        
        while(j < n) {
            mp[fruits[j]]++;
            while(mp.size() > 2) {
                mp[fruits[i]]--;
                if(mp[fruits[i]] == 0) mp.erase(fruits[i]);
                i++;
            }
            if(mp.size() <= 2) res = max(j - i + 1, res);
            j++;
        }
        
        return res;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int i = 0, j = 0, n = fruits.length, maxlen = 0;
        // Map to track count of each fruit type in current window
        Map<Integer, Integer> mp = new HashMap<>();

        // Expand window by moving right pointer
        while (j < n) {
            // Add fruit at j to the window
            mp.put(fruits[j], mp.getOrDefault(fruits[j], 0) + 1);

            // Shrink window from left until we have at most 2 fruit types
            while (mp.size() > 2) {
                // Remove one instance of the leftmost fruit
                mp.put(fruits[i], mp.get(fruits[i]) - 1);
                // If no more of this fruit in window, remove it from map
                if (mp.get(fruits[i]) == 0) {
                    mp.remove(fruits[i]);
                }
                i++;
            }

            // Update max window size seen so far
            maxlen = Math.max(maxlen, j - i + 1);
            j++;
        }

        return maxlen;
    }
}